src/core/credentials/transport/tls/spiffe_utils.cc

```diff
@@ -0,0 +1,158 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include "src/core/credentials/transport/tls/spiffe_utils.h"
+
+#include <string>
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_split.h"
+
+constexpr absl::string_view SPIFFE_PREFIX = "spiffe://";
+
+namespace grpc_core {
+namespace experimental {
+namespace {
+
+absl::Status doInitialUriValidation(absl::string_view uri) {
+  if (uri.empty()) {
+    return absl::InvalidArgumentError(
+        "SpiffeId cannot be parsed from empty uri");
+  }
+  if (uri.length() > 2048) {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "URI length is %d, maximum allowed for SPIFFE ID is 2048",
+        uri.length()));
+  }
+  if (absl::StrContains(uri, "#")) {
+    return absl::InvalidArgumentError(
+        "SPIFFE ID cannot contain query fragments");
+  }
+  if (absl::StrContains(uri, "?")) {
+    return absl::InvalidArgumentError(
+        "SPIFFE ID cannot contain query parameters");
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validateTrustDomain(absl::string_view trust_domain) {
+  if (trust_domain.empty()) {
+    return absl::InvalidArgumentError("Trust domain cannot be empty");
+  }
+  if (trust_domain.size() >= 256) {
+    return absl::InvalidArgumentError(
+        "Trust domain maximum length is 255 characters");
+  }
+  for (auto c : trust_domain) {
+    if (c >= 'a' && c <= 'z') continue;
+    if (c >= '0' && c <= '9') continue;
+    if (c == '.') continue;
+    if (c == '-') continue;
+    if (c == '_') continue;
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "Trust domain contains invalid character %c. MUST contain only "
+        "lowercase letters, numbers, dots, dashes, and underscores",
+        c));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validatePathSegment(absl::string_view path_segment) {
+  if (path_segment.empty()) {
+    return absl::InvalidArgumentError("Path segment cannot be empty");
+  }
+  if (path_segment == "." || path_segment == "..") {
+    return absl::InvalidArgumentError(
+        "Path segment cannot be a relative modifier (. or ..)");
+  }
+  for (auto c : path_segment) {
+    if (c >= 'a' && c <= 'z') continue;
+    if (c >= 'A' && c <= 'Z') continue;
+    if (c >= '0' && c <= '9') continue;
+    if (c == '.') continue;
+    if (c == '-') continue;
+    if (c == '_') continue;
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "Path segment contains invalid character %c. MUST contain only "
+        "letters, numbers, dots, dashes, and underscores",
+        c));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status validatePath(absl::string_view path) {
+  if (path.empty()) {
+    return absl::OkStatus();
```

Comment:
Is an empty path allowed?

Reply:
Yes: [Spec Section](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md#2-spiffe-identity)

> Valid SPIFFE IDs MUST have the scheme set to spiffe, include a non-zero trust domain name, and MUST NOT include a query or fragment component. In other words, a SPIFFE ID is defined in its entirety by the spiffe scheme and a site-specific hier-part which includes an authority component and an ***optional path***.